onnx/mapping.py

```diff
@@ -2,52 +2,42 @@
-    int(TensorProto.FLOAT16): np.dtype('float16'),
-    int(TensorProto.BFLOAT16): np.dtype('float32'),  # Native numpy does not support bfloat16 so now use float32 for bf16 values
-    int(TensorProto.DOUBLE): np.dtype('float64'),
-    int(TensorProto.COMPLEX64): np.dtype('complex64'),
-    int(TensorProto.COMPLEX128): np.dtype('complex128'),
-    int(TensorProto.UINT32): np.dtype('uint32'),
-    int(TensorProto.UINT64): np.dtype('uint64'),
-    int(TensorProto.STRING): np.dtype('object')
+# tensor_type: (numpy type, storage type, string name)
+TENSOR_TYPE_MAP = {
+    int(TensorProto.FLOAT): (np.dtype('float32'), int(TensorProto.FLOAT), 'TensorProto.FLOAT'),
+    int(TensorProto.UINT8): (np.dtype('uint8'), int(TensorProto.INT32), 'TensorProto.UINT8'),
+    int(TensorProto.INT8): (np.dtype('int8'), int(TensorProto.INT32), 'TensorProto.INT8'),
+    int(TensorProto.UINT16): (np.dtype('uint16'), int(TensorProto.INT32), 'TensorProto.UINT16'),
+    int(TensorProto.INT16): (np.dtype('int16'), int(TensorProto.INT32), 'TensorProto.INT16'),
+    int(TensorProto.INT32): (np.dtype('int32'), int(TensorProto.INT32), 'TensorProto.INT32'),
+    int(TensorProto.INT64): (np.dtype('int64'), int(TensorProto.INT64), 'TensorProto.INT64'),
+    int(TensorProto.BOOL): (np.dtype('bool'), int(TensorProto.INT32), 'TensorProto.BOOL'),
+    int(TensorProto.FLOAT16): (np.dtype('float16'), int(TensorProto.UINT16), 'TensorProto.FLOAT16'),
+    # Native numpy does not support bfloat16 so now use float32 for bf16 values
+    # TODO ONNX should dirtectly use bfloat16 for bf16 values after numpy has supported bfloat16 type
+    int(TensorProto.BFLOAT16): (np.dtype('float32'), int(TensorProto.UINT16), 'TensorProto.BFLOAT16'),
+    int(TensorProto.DOUBLE): (np.dtype('float64'), int(TensorProto.DOUBLE), 'TensorProto.DOUBLE'),
+    int(TensorProto.COMPLEX64): (np.dtype('complex64'), int(TensorProto.FLOAT), 'TensorProto.COMPLEX64'),
+    int(TensorProto.COMPLEX128): (np.dtype('complex128'), int(TensorProto.DOUBLE), 'TensorProto.COMPLEX128'),
+    int(TensorProto.UINT32): (np.dtype('uint32'), int(TensorProto.UINT32), 'TensorProto.UINT32'),
+    int(TensorProto.UINT64): (np.dtype('uint64'), int(TensorProto.UINT64), 'TensorProto.UINT64'),
+    int(TensorProto.STRING): (np.dtype('object'), int(TensorProto.STRING), 'TensorProto.STRING'),
 }
 
+# This map is used for converting TensorProto values into numpy arrays
+TENSOR_TYPE_TO_NP_TYPE = {tensor_type: value[0] for tensor_type, value in TENSOR_TYPE_MAP.items()}
+# This is only used to get keys into STORAGE_TENSOR_TYPE_TO_FIELD.
+# TODO(https://github.com/onnx/onnx/issues/4261): Remove this.
+warnings.warn(str("`mapping.TENSOR_TYPE_TO_STORAGE_TENSOR_TYPE` is deprecated in ONNX 1.13 and will be removed in next release. To silence this warning, please use `to_storage_tensor_type` instead."), DeprecationWarning, stacklevel=2)
+TENSOR_TYPE_TO_STORAGE_TENSOR_TYPE = {tensor_type: value[1] for tensor_type, value in TENSOR_TYPE_MAP.items()}
```

Comment:
Can you figure out a way to have it warn only when the dict is accessed? Maybe something like:

```python
class WarningDict(dict):
   def __getitem__(self, key):
       warnings.warn(...)
       return super().__getitem__(self, key)
```

Reply:
Brilliant! Thanks for sharing. I tried it and it worked well.